src/Juvix/Core/FromFrontend.hs

```diff
@@ -236,50 +335,192 @@ transformUsage (FE.Constant (FE.Number (FE.Integer' i)))
+  Just <$> transformValSig def π msig
+transformSig _ def@(Ctx.Record _ msig) =
+  Just <$> transformValSig def Nothing msig
+transformSig x (Ctx.TypeDeclar typ) = Just <$> transformTypeSig x typ
+transformSig _ (Ctx.Unknown sig) =
   throwFF $ UnknownUnsupported $ FE.signatureName <$> sig
-transformSig Ctx.CurrentNameSpace = _ -- TODO ???
+transformSig _ Ctx.CurrentNameSpace = pure Nothing
+transformSig _ (Ctx.Information {}) = pure Nothing
 
-transformType ::
-  forall primTy primVal.
+transformTypeSig ::
+  NameSymbol.T ->
   FE.Type ->
-  Env primTy primVal (HR.Term primTy primVal)
-transformType (FE.Typ {typeForm = FE.Arrowed {dataArrow}}) =
-  transformTermHR dataArrow
-transformType typ@(FE.Typ {typeForm = FE.NonArrowed {}}) =
-  throwFF $ SigRequired $ Ctx.TypeDeclar typ
+  Env primTy primVal (CoreSigHR primTy primVal)
+transformTypeSig name (FE.Typ {typeArgs, typeForm}) = do
+  baseTy <- case typeForm of
+    FE.Arrowed {dataArrow} -> transformTermHR dataArrow
+    FE.NonArrowed {} -> pure $ HR.Star 0 -- TODO metavar (level might not be 0)
+  pure $
+    DataSig
+      { dataType = foldr makeTPi baseTy typeArgs,
+        dataHead = case typeForm of
+          FE.Arrowed {} -> Nothing
+          FE.NonArrowed {} -> Just $ HR.Elim $ foldl HR.App hd args
+            where
+              hd = HR.Var name
+              args = HR.Elim . HR.Var . NameSymbol.fromSymbol <$> typeArgs
+      }
+  where
+    makeTPi name res =
+      -- TODO metavars for the named args instead of defaulting to types
+      HR.Pi mempty (NameSymbol.fromSymbol name) (HR.Star 0) res
 
 transformValSig ::
   FE.Final Ctx.Definition ->
   Maybe Usage.T ->
   Maybe FE.Signature ->
   Env primTy primVal (CoreSigHR primTy primVal)
-transformValSig _ π (Just (FE.Sig _ π' ty cons))
-  | null cons = ValSig _π <$> transformTermHR ty
-  | otherwise = throwFF $ ConstraintsUnsupported cons
-transformValSig def π Nothing = throwFF $ SigRequired def
+transformValSig _ _ (Just (FE.Sig _ (Just π) ty cons))
+  | null cons = ValSig <$> transformGUsage π <*> transformTermHR ty
+  | otherwise = throwFF $ ConstraintsUnimplemented cons
+transformValSig def _ _ = throwFF $ SigRequired def
 
 transformDef ::
+  (Data primTy, Data primVal) =>
+  NameSymbol.T ->
   FE.Final Ctx.Definition ->
-  Env primTy primVal [IR.Global primTy primVal]
-transformDef (Ctx.Def π _ def _) = _
-transformDef (Ctx.Record rec _) = _
-transformDef (Ctx.TypeDeclar typ) = _
-transformDef (Ctx.Unknown _) = pure []
-transformDef Ctx.CurrentNameSpace = pure [] -- FIXME ???
+  Env primTy primVal [IR.RawGlobal primTy primVal]
+transformDef x (Ctx.Def _ _ def _) = do
+  (π, typ) <- getValSig x
+  clauses <- traverse transformClause def
+  let f =
+        IR.Function
+          { funName = x,
+            funUsage = π,
+            funType = hrToIR typ,
+            funClauses = clauses
+          }
+  pure [IR.GFunction f]
+transformDef _ d@(Ctx.Record _ _) = throwFF $ DefUnimplemented d
```

Comment:
Don't we have an issue where any module name `Bar.Foo` will trigger this def undefined?

Do we secure against this anywhere?

Reply:
(since updated)